File: opencv_opengl/Solutions/hw3_solution_python_dana_nguyen/OpenGL_CV.py
from cv2 import cv2
import numpy as np
from OpenGL.GL import *
from OpenGL.GLU import *
from OpenGL.GLUT import *
import sys
import math
import pickle
import time


class OpenGL_CV:

    def __init__(self, source):
        self.cap = cv2.VideoCapture(source)
        if self.cap.isOpened() == False:
            print("Error opening video stream...exiting")
            sys.exit(0)
        self.width = int(self.cap.get(3))
        self.height = int(self.cap.get(4))
        self.frame = None
        self.flipped_frame = None
        self.exCalCam = None
        self.M = None
        self.corners = None
        self.objp = None
        self.projPoints = None
        self.keyPress = False
        self.objpoints3D = []  # 3d point in real world space
        self.imgpoints2D = []  # 2d points in image plane.
        self.tvec = None
        self.rvec = None
        self.rotM = None
        self.ret = False
        #self.calCam = None
        if len(sys.argv) == 2:
            with open(sys.argv[1], "rb") as input:
                self.calCam = pickle.load(input)
                print("calCam:"+str(calCam))
        #elif len(sys.argv) == 3:
        else:
            print("argv[2]:"+str(sys.argv[2]))
            with open(sys.argv[2], "rb") as input:
                self.calCam = pickle.load(input)
                print("calCam:"+str(self.calCam))
        pass

    def draw_axes(self, length):
        '''a useful function to display your coordinate system'''
        glPushAttrib(GL_POLYGON_BIT | GL_ENABLE_BIT | GL_COLOR_BUFFER_BIT)
        glPolygonMode(GL_FRONT_AND_BACK, GL_LINE)
        glDisable(GL_LIGHTING)

        glBegin(GL_LINES)
        glColor3f(1, 0, 0)
        glVertex3f(0, 0, 0)
        glVertex3f(length, 0, 0)

        glColor3f(0, 1, 0)
        glVertex3f(0, 0, 0)
        glVertex3f(0, length, 0)

        glColor3f(0, 0, 1)
        glVertex3f(0, 0, 0)
        glVertex3f(0, 0, length)
        glEnd()

        glPopAttrib()

    def display(self):
        H_IDX = 0  # calling shape on frame returns tuple, 0th index represents height
        W_IDX = 1  # calling shape on frame returns tuple, 1st index represents width
        # clear the window
        glClear(GL_COLOR_BUFFER_BIT)

        # show the current camera frame
        # based on the way opencv stores data, you need to flip it before displaying it
        self.flipped_frame = cv2.flip(self.frame, 0)
        glDrawPixels(self.flipped_frame.shape[W_IDX], self.flipped_frame.shape[H_IDX],
                     GL_BGR, GL_UNSIGNED_BYTE, self.flipped_frame)

        ########################################################################################################################
        # here, set up new parameters to render a scene viewed from the camera

        # set projection matrix using intrinsic camera params
        glMatrixMode(GL_PROJECTION)
        glLoadIdentity()

        # set perspective matrix
        glLoadMatrixf(self.get_persepective_matrix())

        # set viewport
        glViewport(
            0, 0, self.flipped_frame.shape[W_IDX], self.flipped_frame.shape[H_IDX])

        # you will have to set modelview matrix using extrinsic camera params
        glMatrixMode(GL_MODELVIEW)
        glLoadIdentity()
        if self.ret == True:
            glLoadMatrixf(self.M)

            ########################################################################################################################
            # drawing routine
            # now that the camera params have been set, draw your 3D shapes
            # first, save the current matrix
            glPushMatrix()
            if self.keyPress:
                self.draw_sphere(8, 6)
            else:
                self.draw_teapot(8, 6)
            self.draw_axes(1.0)
            glPopMatrix()

        # show the rendering on the screen
        glutSwapBuffers()

        # post the next redisplay
        glutPostRedisplay()

    def get_persepective_matrix(self):
        # creating perspective matrix: https://blog.noctua-software.com/opencv-opengl-projection-matrix.html
        P = np.zeros(shape=(4, 4), dtype=np.float32)

        fx = self.calCam[1][0][0]
        fy = self.calCam[1][1][1]

        cx = self.calCam[1][0][-1]
        cy = self.calCam[1][1][-1]

        near = 0.1
        far = 100.0

        P[0][0] = 2*fx / self.width
        P[1][1] = 2*fy / self.height
        P[0][2] = 1 - (2*cx / self.width)
        P[1][2] = (2*cy / self.height - 1)
        P[2][2] = -(far + near) / (far - near)
        P[3][2] = -1.
        P[2][3] = -(2*far*near) / (far - near)

        p = P.T
        return p

    def draw_sphere(self, w, h):
        for x in range(w):
            for y in range(h):
                glPushMatrix()
                glTranslatef(x, y, 0)
                glutSolidSphere(0.2, 100, 100)
                glPopMatrix()

    def draw_teapot(self, w, h):
        glPushMatrix()
        glTranslatef(w/2.0, h/2.0, 0)
        glRotatef(270, 1, 0, 0)
        glutSolidTeapot(1.5)
        glPopMatrix()

    def reshape(self, w, h):
        '''set openGL viewport (drawable area)'''
        glViewport(0, 0, w, h)

    def mouse(self, button, state, x, y):
        if button == GLUT_LEFT_BUTTON and state == GLUT_UP:
            pass

    def keyboard(self, key, x, y):
        if key == b'q':
            # quit when 'q' is pressed
            sys.exit(0)
        if key == b' ':
            if self.keyPress == False:
                self.keyPress = True
            elif self.keyPress == True:
                self.keyPress = False
        else:
            pass

    def draw_debug_circles(self):
        # used to debug on checkerboard
        imgpoints2D2, _ = cv2.projectPoints(
            self.objp, self.rvec, self.tvec, self.calCam[1], self.calCam[2])

        for i in imgpoints2D2:
            coordinate = (i[0][0], i[0][1])
            cv2.circle(self.frame, coordinate, 5, color=(255, 0, 0))

    def idle(self):
        '''grabs a frame from the camera'''
        ret_frame, frame = self.cap.read()

        if ret_frame == False:
            print('no frames to grab, exiting')
            sys.exit(0)

        self.frame = frame

        # used to slow down the video
        # time.sleep(1)

        # storing object points and image points for frame
        objp = np.zeros((6*8, 3), np.float32)
        objp[:, :2] = np.mgrid[0:8, 0:6].T.reshape(-1, 2)
        self.objp = objp

        # finding corners of the frame
        ret, corners = cv2.findChessboardCorners(frame, (8, 6), None)
        self.corners = corners
        self.ret = ret

        if ret == True:
            self.objpoints3D.append(objp)
            self.imgpoints2D.append(corners)
            # getting model view matrix
            # https://stackoverflow.com/questions/44375149/opencv-to-opengl-coordinate-system-transform
            ret_, rvec, tvec = cv2.solvePnP(
                objp, corners, self.calCam[1], self.calCam[2])

            # print(ret_, rvec, tvec)
            self.tvec = tvec
            self.rvec = rvec
            rotM = cv2.Rodrigues(rvec)[0]
            M = [[], [], [], [0, 0, 0, 1]]
            self.rotM = rotM
            for idx, val in enumerate(rotM):
                # invert y and z axis
                if idx == 1 or idx == 2:
                    M[idx] = np.multiply(
                        np.array(list(rotM[idx]) + [tvec[idx][0]]), -1)
                    # print(np.dot(rotM[idx], Rx[idx]))
                else:
                    M[idx] = np.array(
                        list(rotM[idx])+[tvec[idx][0]])
            self.M = np.array(M).T


def main(argv):
    if len(argv) == 2:
        source = 0
    else:
        source = argv[1]

    print("source:"+str(source))
    ogl_cv = OpenGL_CV(source)
    ogl_cv.idle()

    # initialize GLUT
    glutInit()
    glutInitDisplayMode(GLUT_RGBA | GLUT_DOUBLE)
    # glutInitWindowPosition(20, 20)
    glutInitWindowSize(ogl_cv.width, ogl_cv.height)
    glutCreateWindow("OpenGL / OpenCV Example")

    # set up GUI callback functions
    glutDisplayFunc(ogl_cv.display)
    glutReshapeFunc(ogl_cv.reshape)
    glutMouseFunc(ogl_cv.mouse)
    glutKeyboardFunc(ogl_cv.keyboard)
    glutIdleFunc(ogl_cv.idle)

    # start GUI loop
    glutMainLoop()


if __name__ == '__main__':
    main(sys.argv)

File: opencv_opengl/Solutions/hw3_solution_cpp/opengl/opengl_cv.cpp
// Skeleton Code for CS290I Homework 1
// 2012, Jon Ventura and Chris Sweeney

// adapt the include statements for your system:

#include <opencv2/opencv.hpp>
#include <opencv2/highgui/highgui_c.h>
#include <chrono>
#include <thread>

#ifdef __APPLE__
#include <GLUT/glut.h>
#else
#include <GL/gl.h>
#include <GL/glut.h>
#endif

#include <cstdio>
#include <vector>
#include <math.h>

// IF DEBUG is defined, print tons of debug output.
#ifdef DEBUG
static bool debug = true;
#else
static bool debug = false;
#endif

// Must match the dimensions of the filmed calibration pattern.
#define GRID_WIDTH 8
#define GRID_HEIGHT 6
#define DISTORTION_CORRECTION 0
#define SLOWDOWN 0

using namespace std;

cv::VideoCapture *cap = NULL;
int width = 640;
int height = 480;
cv::Mat image;
cv::Mat distCoeffs;
int frame_counter = 0;
int g_nr_frames = 0;

// Intrinsic camera parameters.
static float fovx, fovy, focalLength, principalX, principalY, fx, fy, k1, k2, p1, p2;
cv::Mat cameraMat;

// Current AR overlay.
int overlay = 0;

// a useful function for displaying your coordinate system
void drawAxes(float length)
{
  //  return;  
  glPushAttrib(GL_POLYGON_BIT | GL_ENABLE_BIT | GL_COLOR_BUFFER_BIT) ;

  glPolygonMode(GL_FRONT_AND_BACK, GL_LINE) ;
  glDisable(GL_LIGHTING) ;

  glBegin(GL_LINES) ;
  glColor3f(1,0,0) ;
  glVertex3f(0,0,0) ;
  glVertex3f(length,0,0);

  glColor3f(0,1,0) ;
  glVertex3f(0,0,0) ;
  glVertex3f(0,length,0);

  glColor3f(0,0,1) ;
  glVertex3f(0,0,0) ;
  glVertex3f(0,0,length);
  glEnd() ;


  glPopAttrib() ;
}


bool getImagePoints(cv::Mat& image, int width, int height, vector<cv::Point2f>& corners) {
  cv::Size patternSize(width, height);
  bool boardFound = cv::findChessboardCorners(image, patternSize, corners, 0);
  return boardFound;
}


void createObjectPoints(int width, int height, vector<cv::Point3f>& points) {
  for (int y = 0; y < height; y++) {
    for (int x = 0; x < width; x++) {
      points.push_back(cv::Point3f(x, y, 0));
    }
  }
}
// (0,0),(1,0),(2,0)...(9,0),(0,1)...(9,1)(0,2)... ...(9,6)

void display()
{
  if (SLOWDOWN) std::this_thread::sleep_for(std::chrono::milliseconds(SLOWDOWN));

  if (!image.data) {
    // This randomly occurs at the beginning of a video file, and always at its end.
    glutPostRedisplay();
    return;
  }

  int width = image.size().width;
  int height = image.size().height;

  cv::Mat tempimage;
  cv::flip(image, tempimage, 0);
  glViewport(0, 0, tempimage.size().width, tempimage.size().height);

  glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

  // Set projection matrix using intrinsic camera params.
  glMatrixMode(GL_PROJECTION);
  glLoadIdentity();
  glFrustum(-principalX / fx, (width - principalX) / fy, (principalY - height) / fy, principalY / fy, 1, 500);  
  
  // Locate calibration pattern.
  vector<cv::Point3f> objectPoints;
  vector<cv::Point2f> imagePoints;
  createObjectPoints(GRID_WIDTH, GRID_HEIGHT, objectPoints);
  bool found = getImagePoints(image, GRID_WIDTH, GRID_HEIGHT, imagePoints);
  if (!found) {
    // We couldn't find the (entire) pattern, so we can't determine the extrinsic parameters for this frame.
    printf("%d points not found \n", GRID_WIDTH * GRID_HEIGHT - imagePoints.size());
    cv::flip(image, tempimage, 0);
    glDisable(GL_DEPTH_TEST);
    glDrawPixels( tempimage.size().width, tempimage.size().height, GL_BGR, GL_UNSIGNED_BYTE, tempimage.ptr() );
    glEnable(GL_DEPTH_TEST);
    glutSwapBuffers();
    glutPostRedisplay();
    return;
  }

  // Solve for extrinsic parameters.
  cv::Mat rvec, tvec;
  cv::solvePnP(objectPoints, imagePoints, cameraMat, distCoeffs, rvec, tvec);
  cv::Mat undistorted;
  cv::undistort(image, undistorted, cameraMat, distCoeffs);

  // Draw points projected with OpenCV. Useful for verifying intrinsic/extrinsic parameters 
  // and debugging the OpenGL projection matrix.
  if (debug) {
    vector<cv::Point2f> projected;
    cv::projectPoints(objectPoints, rvec, tvec, cameraMat, distCoeffs, projected);
    for (int i = 0; i < projected.size(); i++) {
      cv::Point2f pt = projected.at(i);
      cv::circle(image, pt, 5, cv::Scalar(0, 0, 255));
    }
  }

  if (DISTORTION_CORRECTION)
    {
      // use undistorted camera image: 
      cv::flip(undistorted, tempimage, 0);
    } else 
    {
      // use original camera image:
      cv::flip(image, tempimage, 0);
    }
    
  glDisable(GL_DEPTH_TEST);
  glDrawPixels(tempimage.size().width, tempimage.size().height, GL_BGR, GL_UNSIGNED_BYTE, tempimage.ptr());
  glEnable(GL_DEPTH_TEST);

  // Convert rvec, tvec (from solvePnP) to transformation matrix.
  cv::Mat rm = cv::Mat(3, 3, CV_64F);
  cv::Rodrigues(rvec, rm); // Convert rotation vector to rotation matrix.
  // Top-left 3x3 = the rotation matrix. Top-right 3x1 = translation vector.
  const double xform[16] = {
    rm.at<double>(0,0), rm.at<double>(0,1), rm.at<double>(0,2), tvec.at<double>(0),
    rm.at<double>(1,0), rm.at<double>(1,1), rm.at<double>(1,2), tvec.at<double>(1),
    rm.at<double>(2,0), rm.at<double>(2,1), rm.at<double>(2,2), tvec.at<double>(2),
    0                 , 0                 , 0                 , 1
  };
  
  // Print the transformation matrix.
  if (debug) {
    printf("------------------------\n");
    for(int y = 0; y < 4; y++) {
      printf("|%f\t%f\t%f\t%f|\n", xform[4*y+0],xform[4*y+1],xform[4*y+2],xform[4*y+3]);
    }
    printf("------------------------\n");
  }    
 
  glMatrixMode(GL_MODELVIEW);
  glLoadIdentity();
  glScalef(1.0,-1.0, -1.0); // Welcome to the other hand...

  glMultTransposeMatrixd(xform); // Transpose, because we constucted xform row-major above.
 

  // Draw everything.
  float lightPos[] = {50, 50, -50, 0};
  glLightfv(GL_LIGHT0, GL_POSITION, lightPos);

  drawAxes(1.0);

  switch (overlay) {
  case 0:
    glPushMatrix();
    if (debug) {
      // Draw a ground plane to help position the teapot.
      glDisable(GL_LIGHTING);
      glBegin(GL_QUADS);
      glVertex3f(10,10,0);
      glVertex3f(0,10,0);
      glVertex3f(0,0,0);
      glVertex3f(10,0,0);
      glEnd();
      glEnable(GL_LIGHTING);
    }
    glTranslatef(GRID_WIDTH/2, GRID_HEIGHT/2, -1.5);
    glRotatef(-90, 1, 0, 0);
    glutSolidTeapot(1.5);
    glPopMatrix();
    break;

  case 1:
    for(int i = 0; i < objectPoints.size(); i++) {
      cv::Point3f pt = objectPoints.at(i);
      glPushMatrix();
      glTranslatef(pt.x, pt.y, 0);
      glutSolidSphere(.3, 10, 10);
      glPopMatrix();
    }
    break;
  }

  // show the rendering on the screen
  glutSwapBuffers();

  // post the next redisplay
  glutPostRedisplay();
  
}

void reshape(int w, int h) {
  // set OpenGL viewport (drawable area)
  glViewport( 0, 0, w, h );
}

void mouse(int button, int state, int x, int y) {
}

void keyboard(unsigned char key, int x, int y) {
  switch ( key )
    {
    case 'q':
      // quit when q is pressed
      exit(0);
      break;

    case ' ':
      // Press [space] to alternate between overlays.
      overlay ^= 1;
      break;

    default:
      break;
    }
}

void idle() {
  // grab a frame from the camera
  (*cap) >> image;

  if (++frame_counter == g_nr_frames-1)
    {
      frame_counter = 0;
      cap->set(cv::CAP_PROP_POS_FRAMES, 0);
    }

}

// MAIN Function
////////////////

int main(int argc, char **argv) {
  int w,h;

  FILE *params = stdin;
  // Or, from a file.
  // FILE *params = fopen("params.txt", "r");

  if ( argc == 1 ) {
    // start video capture from camera
    cap = new cv::VideoCapture(0);
  } else if ( argc == 2 ) {
    // start video capture from file
    printf("From file\n");
    cap = new cv::VideoCapture(argv[1]);
  } else {
    fprintf( stderr, "usage: %s [<filename>]\n", argv[0] );
    return 1;
  }

  // check that video is opened
  if ( cap == NULL || !cap->isOpened() ) {
    fprintf( stderr, "could not start video capture\n" );
    return 1;
  }

  g_nr_frames = cap->get(cv::CAP_PROP_FRAME_COUNT);
    
  printf("Reading intrinsic parameters from STDIN...\n");

  // Load intrinsic parameters.
  fscanf(params, "%f\n", &fovx);
  fscanf(params, "%f\n", &fovy);
  fscanf(params, "%f\n", &focalLength);
  fscanf(params, "%f\n", &principalX);
  fscanf(params, "%f\n", &principalY);
  fscanf(params, "%f\n", &fx);
  fscanf(params, "%f\n", &fy);
  fscanf(params, "%f\n", &k1);
  fscanf(params, "%f\n", &k2);
  fscanf(params, "%f\n", &p1);
  fscanf(params, "%f\n", &p2);

  // Create a camera matrix.
  float m[] = {fx, 0, principalX, 
	       0, fy, principalY, 
	       0, 0, 1};
  cameraMat = cv::Mat(3, 3, CV_32F, m);

  // And the distortion coefficients matrix.
  float d[] = {k1, k2, p1, p2};
  distCoeffs = cv::Mat(1, 4, CV_32F, d);

  // get width and height
  w = (int) cap->get( cv::CAP_PROP_FRAME_WIDTH );
  h = (int) cap->get( cv::CAP_PROP_FRAME_HEIGHT );
  // On Linux, there is currently a bug in OpenCV that returns 
  // zero for both width and height here (at least for video from file)
  // hence the following override to global variable defaults: 
  width = w ? w : width;
  height = h ? h : height;

  // initialize GLUT
  glutInit(&argc, argv);
  glutInitDisplayMode(GLUT_DEPTH | GLUT_RGB | GLUT_DOUBLE);
  glutInitWindowPosition(20, 20);
  glutInitWindowSize(width, height);
  glutCreateWindow("CS290I Assignment 1");

  // Initialize GL.
  glHint(GL_PERSPECTIVE_CORRECTION_HINT, GL_NICEST);
  glDepthFunc(GL_LESS);
  glShadeModel(GL_FLAT);
  glDisable(GL_CULL_FACE); // Nicer looking teapots.
  glClearColor(0.0f, 0.0f, 0.0f, 1);
  glEnable(GL_DEPTH_TEST);
  glClearDepth(1.0f);
  glEnable(GL_LIGHTING);
  glEnable(GL_LIGHT0);

  // set up GUI callback functions
  glutDisplayFunc( display );
  glutReshapeFunc( reshape );
  glutMouseFunc( mouse );
  glutKeyboardFunc( keyboard );
  glutIdleFunc( idle );

  // start GUI loop
  glutMainLoop();

  return 0;
}
